packages/@lwc/engine-dom/src/styles.ts

```diff
@@ -0,0 +1,231 @@
+        elm = createFreshStyleElement(content);
+        styleElements.set(content, elm);
+    } else {
+        // This `<style>` may be repeated multiple times in the DOM, so cache it. It's a bit
+        // faster to call `cloneNode()` on an existing node than to recreate it every time.
+        elm = elm.cloneNode(true) as HTMLStyleElement;
+    }
+    return elm;
+}
+
+function setStyleElementForTarget(
+    target: ShadowRoot | Document,
+    content: string,
+    elm: HTMLStyleElement
+) {
+    let contentsToStyleElements = targetsToStyleElements.get(target);
+    if (isUndefined(contentsToStyleElements)) {
+        contentsToStyleElements = new Map();
+        targetsToStyleElements.set(target, contentsToStyleElements);
+    }
+    contentsToStyleElements.set(content, elm);
+}
+
+function unsetStyleElementForTarget(target: ShadowRoot | Document, content: string) {
+    const contentsToStyleElements = targetsToStyleElements.get(target);
+    if (!isUndefined(contentsToStyleElements)) {
+        contentsToStyleElements.delete(content);
+    }
+}
+
+function getStyleElementForTarget(target: ShadowRoot | Document, content: string) {
+    const contentsToStyleElements = targetsToStyleElements.get(target);
+    if (!isUndefined(contentsToStyleElements)) {
+        return contentsToStyleElements.get(content);
+    }
+}
+
+function createOrGetConstructableStyleSheet(content: string) {
+    // It's important for CSSStyleSheets to be unique based on their content, so
+    // that adoptedStyleSheets.indexOf(sheet) works
+    let styleSheet = styleSheets.get(content);
+    if (isUndefined(styleSheet)) {
+        styleSheet = new CSSStyleSheet();
+        styleSheet.replaceSync(content);
+        styleSheets.set(content, styleSheet);
+    }
+    return styleSheet;
+}
+
+function getStylesToUsageCounts(target: ShadowRoot | Document) {
+    let stylesToCounts = stylesToUsageCount.get(target);
+    if (isUndefined(stylesToCounts)) {
+        stylesToCounts = new Map();
+        stylesToUsageCount.set(target, stylesToCounts);
+    }
+    return stylesToCounts;
+}
+
+function incrementOrDecrementUsageCount(
+    content: string,
+    target: ShadowRoot | Document,
+    delta: number
+) {
+    const stylesToCounts = getStylesToUsageCounts(target);
+
+    let count = stylesToCounts.get(content);
+    if (isUndefined(count)) {
+        count = 0;
+    }
+    count += delta;
+    stylesToCounts.set(content, count);
+    return count;
+}
+
+function insertConstructableStyleSheet(content: string, target: ShadowRoot | Document) {
+    const styleSheet = createOrGetConstructableStyleSheet(content);
+    const { adoptedStyleSheets } = target;
+    if (supportsMutableAdoptedStyleSheets) {
+        // This is only supported in later versions of Chromium:
+        // https://chromestatus.com/feature/5638996492288000
+        adoptedStyleSheets.push(styleSheet);
+    } else {
+        target.adoptedStyleSheets = [...adoptedStyleSheets, styleSheet];
+    }
```

Comment:
Previously we were _not_ using constructable stylesheets for the global `document`. The main reason was performance (there are a lot of stylesheets on the `document` compared to a single shadow root). But now that we have mutable adopted stylesheets shipping in Chrome 99, there shouldn't be a perf penalty for appending to the `adoptedStyleSheets` array.